### PR TITLE
LANG-1480 getAbbreviatedName refactored to create appropriate length …

### PR DESCRIPTION
**Pull request by verhas**
…short class names

### CODE REVIEW

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 1]
```diff
@@ -16,19 +16,11 @@
...
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
-
-import org.apache.commons.lang3.mutable.MutableObject;
+import java.util.*;
```
Comment (kinow) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: This won't pass checkstyle I think. You will need to replace the `*` by explicit imports.

See Travis builds failing, this part is due to this change: 

>[ERROR] src/main/java/org/apache/commons/lang3/ClassUtils.java:[23] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - java.util.*.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java [hunk 2]
```diff
@@ -178,8 +179,22 @@ public void test_getAbbreviatedName_Class_ZeroLen() {
...
+        assertEquals(".", ClassUtils.getAbbreviatedName(".", 18));
+        assertEquals(".", ClassUtils.getAbbreviatedName(".", 1));
+        assertEquals("..", ClassUtils.getAbbreviatedName("..", 1));
+        assertEquals("...", ClassUtils.getAbbreviatedName("...", 2));
+        assertEquals("...", ClassUtils.getAbbreviatedName("...", 3));
+        assertEquals("java.lang.String", ClassUtils.getAbbreviatedName("java.lang.String", Integer.MAX_VALUE));
+        assertEquals("j.lang.String", ClassUtils.getAbbreviatedName("java.lang.String", "j.lang.String".length()));
+        assertEquals("j.l.String", ClassUtils.getAbbreviatedName("java.lang.String", "j.lang.String".length() - 1));
+        assertEquals("j.l.String", ClassUtils.getAbbreviatedName("java.lang.String", "j.l.String".length()));
+        assertEquals("j.l.String", ClassUtils.getAbbreviatedName("java.lang.String", "j.l.String".length() - 1));
```
Comment (kinow) [inline src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: I think you are not testing the bug you found? There is no test for a call to `getAbbreviatedName` with class name as a `String`, and length as the exact length of that `String`.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 3]
```diff
@@ -433,11 +432,13 @@ public static String getAbbreviatedName(final Class<?> cls, final int len) {
...
      * <tr><td>"java.lang.String"</td><td>30</td><td>"java.lang.String"</td></tr>
+     * <tr><td>"org.apache.commons.lang3.ClassUtils"</td><td>18</td><td>"o.a.c.l.ClassUtils"</td></tr>
      * </table>
      * @param className  the className to get the abbreviated name for, may be {@code null}
      * @param len  the desired length of the abbreviated name
-     * @return the abbreviated name or an empty string
-     * @throws IllegalArgumentException if len &lt;= 0
+     * @return the abbreviated name or an empty string if the specified
+     * class name is {@code null} or empty string
+     * @throws IllegalArgumentException if {@code len <= 0}
```
Comment (kinow) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: Thanks for updating the docs!

**Comment by kinow** [comment]
I think you need to include tests for this change. See the bot comment about coverage decreasing, and a comment I've added about the missing test for the bug you described here.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 1 above)
Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: This was the default IntelliJ setting to use * after more than 4 imports from the same package.

I see no reason why not to use '*' in imports these days with such advanced editors, but in case that is the project standard, I comply. Thanks.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 4]
```diff
@@ -987,33 +1005,35 @@ public static boolean isAssignable(Class<?> cls, final Class<?> toClass, final b
...
-     * @param className  the class name
+     * @param classLoader the class loader to use to load the class
+     * @param className   the class name
      * @param initialize  whether the class must be initialized
      * @return the class represented by {@code className} using the {@code classLoader}
      * @throws ClassNotFoundException if the class is not found
      */
     public static Class<?> getClass(
-            final ClassLoader classLoader, final String className, final boolean initialize) throws ClassNotFoundException {
+        final ClassLoader classLoader, final String className, final boolean initialize) throws ClassNotFoundException {
```
Comment (kinow) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: I think now the PR has a lot more changes that are not related to LANG-1480. Are you able to undo formatting changes, and leave only the changes for the issue, please?

Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: I am doing it. Actually, these formattings are clumsy in the original code, and IntelliJ just reformatted automatically. I understand that it is no excuse for creating a messy PR. Sorry.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 5]
```diff
@@ -433,48 +440,82 @@ public static String getAbbreviatedName(final Class<?> cls, final int len) {
...
+     * <tr><td>"org.apache.commons.lang3.ClassUtils"</td><td>18</td><td>"o.a.c.l.ClassUtils"</td></tr>
      * </table>
-     * @param className  the className to get the abbreviated name for, may be {@code null}
-     * @param len  the desired length of the abbreviated name
-     * @return the abbreviated name or an empty string
-     * @throws IllegalArgumentException if len &lt;= 0
+     *
+     * @param className the className to get the abbreviated name for, may be {@code null}
+     * @param len       the desired length of the abbreviated name
+     * @return the abbreviated name or an empty string if the specified
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: Please specify if the result string is always <= len.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 6]
```diff
@@ -424,7 +424,14 @@ public static String getAbbreviatedName(final Class<?> cls, final int len) {
...
      * <p>The abbreviated class name will always include the complete package hierarchy.
      * If enough space is available, rightmost sub-packages will be displayed in full
-     * length.</p>
+     * length. The abbreviated package names will be shortened to a single character.</p>
+     * <p>Only package names are shortened, the class simple name remains untouched. (See examples.)</p>
+     * <p>The result will be longer than the desired length only if all the package names
+     * shortened to a single character plus the class simple name with the separating dots
+     * together are longer than the desired length. In other words, when the class name
+     * cannot be shortened to the desired length without significant loss of meaning.</p>
+     * <p>If the class name can be shortened without significant loss of meaning then
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: I do not think we should talk about "loss of meaning" as to me, abbreviating implies "loss of meaning" by design ;-) Just describe what it does. IMO a list would be better to describe the different use cases.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 7]
```diff
@@ -433,48 +440,82 @@ public static String getAbbreviatedName(final Class<?> cls, final int len) {
...
+     *                       character copied when the current part was
+     *                       copied
+     * @param originalLength the original length of the class full name,
+     *                       which is abbreviated
+     * @param desiredLength  the desired length of the abbreviated class
+     *                       name
+     * @return {@code true} if it can be kept in its original length
+     * {@code false} if the current part has to be abbreviated and
+     */
+    private static boolean useFull(final int runAheadTarget,
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: Is this only used once? If so, just in-line it with a `//` comment.

Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: Is this a strong requirement or something you can live with if I leave it as it is?

As you can see from the length of the comment this is a complex expression and the JavaDoc explains a lot about the algorithm used. Such an extensive commend would degrade the readability of the caller method.

As per the performance: the JIT will inline this method, it should not be a problem.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 5 above)
Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: It is not and it is described in details in the above text.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 7 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: I'm ok to leave as it since the description is nice especially using `@param` tags which would end up in some non-standard format if documented with `//`s.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java [hunk 8]
```diff
@@ -47,7 +47,7 @@
  * Unit tests {@link org.apache.commons.lang3.ClassUtils}.
  */
 @SuppressWarnings("boxing") // JUnit4 does not support primitive equality testing apart from long
-public class ClassUtilsTest  {
+class ClassUtilsTest  {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: Let's not change the visibility of test classes please.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 9]
```diff
@@ -433,48 +440,83 @@ public static String getAbbreviatedName(final Class<?> cls, final int len) {
...
-        } else {
-          if (availableSpace > 0) {
-            output[level] = part;
-          } else {
-            // if no space is left still the first char is used
-            output[level] = part.substring(0, 1);
-          }
+        if (className == null) {
+            return StringUtils.EMPTY;
         }
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: I was expecting a short-circuit here when the length hint is greater or equal than the input string, then return the input string to avoid creating a new String.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java [hunk 10]
```diff
@@ -433,48 +440,83 @@ public static String getAbbreviatedName(final Class<?> cls, final int len) {
...
+     *
+     * @param className the className to get the abbreviated name for, may be {@code null}
+     * @param len       the desired length of the abbreviated name
+     * @return the abbreviated name or an empty string if the specified
+     * class name is {@code null} or empty string. The abbreviated name may be
+     * longer than the desired length if it cannot be abbreviated to the desired length.
+     * @throws IllegalArgumentException if {@code len <= 0}
      * @since 3.4
      */
     public static String getAbbreviatedName(final String className, final int len) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: I am starting to wonder if `len` would be better named `lengthHint` to make it clearer that this length is not a guarantee in any shape or form of the length of the returned string.

Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: I can live with that. Makes sense.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 9 above)
Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: Makes sense.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 10 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: @verhas Then please go ahead and change it to `lengthHint`.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 9 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: @verhas Then please go ahead and update the code and add a test.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java (hunk 8 above)
Comment (garydgregory) [reply src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: @verhas ping.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 9 above)
Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: There is no point to add a test. This is not a feature change. This is a performance optimization. I do not think the method should guarantee as an API contract that it will return the same String object in this case.

In other cases I totally agree: if there is a change, which is functionality change then there should be a test also changed.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java (hunk 8 above)
Comment (garydgregory) [reply src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: @verhas ping.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java [hunk 11]
```diff
@@ -160,26 +162,59 @@ public void test_getAbbreviatedName_Class() {
...
         assertEquals("java.lang.String", ClassUtils.getAbbreviatedName(String.class, 20));
     }
 
+    /**
+     * Test that in case the required length is larger than the name and thus there is no need for any shortening
+     * then the returned string object is the same as the one passed as argument. Note, however, that this is
+     * tested as an internal implementation detail, but it is not a guaranteed feature of the implementation.
+     */
     @Test
-    public void test_getAbbreviatedName_Class_NegativeLen() {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: Let's not change the visibility of test classes and methods please.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java [hunk 12]
```diff
@@ -160,26 +162,59 @@ public void test_getAbbreviatedName_Class() {
...
+    }
+
+    @Test
+    @DisplayName("When the desired length is negative then exception is thrown")
+    void test_getAbbreviatedName_Class_NegativeLen() {
         assertThrows(IllegalArgumentException.class, () -> ClassUtils.getAbbreviatedName(String.class, -10));
     }
 
     @Test
-    public void test_getAbbreviatedName_Class_ZeroLen() {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: Let's not change the visibility of test classes and methods please.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java [hunk 13]
```diff
@@ -160,26 +162,59 @@ public void test_getAbbreviatedName_Class() {
...
+    @DisplayName("When the desired length is zero then exception is thrown")
+    void test_getAbbreviatedName_Class_ZeroLen() {
         assertThrows(IllegalArgumentException.class, () -> ClassUtils.getAbbreviatedName(String.class, 0));
     }
 
     @Test
-    public void test_getAbbreviatedName_String() {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: Let's not change the visibility of test classes and methods please.

Comment (verhas) [reply src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: Why?

Since the project uses JUnit 5 the required visibility is package private. Any access modifier is noise.

**Comment by garydgregory** [comment]
Because all other tests are public. IOW, consistency.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java (hunk 12 above)
Comment (verhas) [reply src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: Sorry, but wrong addressee. This is not me, who has changed the visibility, but JUnit 5.

File: src/test/java/org/apache/commons/lang3/ClassUtilsTest.java (hunk 13 above)
Comment (verhas) [reply src/test/java/org/apache/commons/lang3/ClassUtilsTest.java]: Sorry, but wrong addressee. This is not me, who has changed the visibility, but JUnit 5.

**Comment by garydgregory** [comment]
> Please merge it instead of nitpicking.

That's not the best attitude IMO.

**Comment by verhas** [comment]
>>Please merge it instead of nitpicking.
>That's not the best attitude IMO.

I said please. Btw: I opened a ticket https://issues.apache.org/jira/browse/LANG-1491 to fix this issue.

**Comment by garydgregory** [comment]
> > > Please merge it instead of nitpicking.
> > > That's not the best attitude IMO.
> 
> I said please. Btw: I opened a ticket https://issues.apache.org/jira/browse/LANG-1491 to fix this issue.

It's rather simple IMO: this patch should be consistent with the current code base as it is now. IOW, tests are `public`.

If someone wants to do the busy work of changing all of Commons Lang, more power to them, then the community can decide what to do, but this PR is not the place to do it.

**Comment by verhas** [comment]
I changed the test class and methods to be public.

**Comment by garydgregory** [comment]
@verhas 
GitHub seems to think there conflicts here. May you rebase?

File: src/main/java/org/apache/commons/lang3/ClassUtils.java (hunk 7 above)
Comment (verhas) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: Thanks

**Comment by garydgregory** [comment]
@verhas 
In git master; please verify and close.
